Testing in Docker environment fails with pex FileNotFoundError

**Describe the bug**
When trying to run some tests in my `docker_environment` I get the below error:

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/runpy.py", line 187, in _run_module_as_main
    mod_name, mod_spec, code = _get_main_module_details(_Error)
  File "/usr/local/lib/python3.8/runpy.py", line 221, in _get_main_module_details
    return _get_module_details(main_name)
  File "/usr/local/lib/python3.8/runpy.py", line 155, in _get_module_details
    code = loader.get_code(mod_name)
  File "<frozen zipimport>", line 159, in get_code
  File "<frozen zipimport>", line 709, in _get_module_code
  File "<frozen zipimport>", line 536, in _get_data
FileNotFoundError: [Errno 2] No such file or directory: './pex'
```

I have included the full traceback below. If you need any more information please let me know.

**Pants version**
2.16.0.dev5

**OS**
I am developing on a Mac but the tests are supposed to be executed in the `python:3.8-slim` linux Docker container.

**Additional info**
In my root level `BUILD` file I have configured the `docker_environment`
```python
docker_environment(
    name="dev_docker",
    platform="linux_x86_64",
    image="python:3.8-slim",
    python_bootstrap_search_path=["<PATH>"],
)
```

The test `BUILD` file has set the environment

```python
python_tests(
    name="tests", environment="dev_docker"
)
```

<details>
  <summary>pants.toml</summary>

```toml
[GLOBAL]
pants_version = "2.16.0.dev5"
dynamic_ui = true
backend_packages = [
  "pants.backend.build_files.fix.deprecations",
  "pants.backend.build_files.fmt.black",
  "pants.backend.python",
  "pants.backend.python.lint.autoflake",
  "pants.backend.python.lint.black",
  "pants.backend.python.lint.flake8",
  "pants.backend.python.lint.isort",
  "pants.backend.python.lint.pyupgrade",
  "pants.backend.python.typecheck.mypy",
  "pants.backend.docker",
]

[python]
enable_resolves = true
interpreter_constraints = [">=3.8.*"]

[python.resolves_to_only_binary]
__default__= [":all:"]

[python-bootstrap]
search_path = [
  "<PYENV>",
  "<PATH>"
]

[test]
batch_size = 16

[pytest]
args = ["--no-header"]
execution_slot_var = "PANTS_RUNNER_ID"
config = "pyproject.toml"

[docker]
tailor = false

[environments-preview.names]
dev_docker = "//:dev_docker"
```
</details>

<details>
  <summary>full error traceback</summary>

```
pants test src/project:: --print-stacktrace --keep-sandboxes=on_failure
13:23:53.31 [INFO] Completed: Pulling Docker image `python:3.8-slim` because the image is missing locally.
13:23:54.77 [INFO] Preserving local process execution dir /private/var/folders/8b/0w6lmjjx7fsdm_6hhcrjfdfw0000gq/T/pants-sandbox-oEZLFk for Find interpreter for constraints: CPython>=3.8.*
13:23:54.78 [ERROR] 1 Exception encountered:

Engine traceback:
  in select
    ..
  in pants.core.goals.test.run_tests
    `test` goal

Traceback (most recent call last):
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 626, in native_engine_generator_send
    res = rule.send(arg) if err is None else rule.throw(throw or err)
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/core/goals/test.py", line 836, in run_tests
    results = await MultiGet(
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 360, in MultiGet
    return await _MultiGet(tuple(__arg0))
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 167, in __await__
    result = yield self.gets
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 626, in native_engine_generator_send
    res = rule.send(arg) if err is None else rule.throw(throw or err)
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/backend/python/goals/pytest_runner.py", line 475, in run_python_tests
    setup = await Get(
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 120, in __await__
    result = yield self
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 626, in native_engine_generator_send
    res = rule.send(arg) if err is None else rule.throw(throw or err)
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/backend/python/goals/pytest_runner.py", line 475, in run_python_tests
    setup = await Get(
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 120, in __await__
    result = yield self
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 626, in native_engine_generator_send
    res = rule.send(arg) if err is None else rule.throw(throw or err)
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/backend/python/goals/pytest_runner.py", line 262, in setup_pytest_for_target
    ) = await MultiGet(
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 451, in MultiGet
    return await _MultiGet((__arg0, __arg1, __arg2, __arg3, __arg4, __arg5))
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 167, in __await__
    result = yield self.gets
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 626, in native_engine_generator_send
    res = rule.send(arg) if err is None else rule.throw(throw or err)
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/backend/python/goals/pytest_runner.py", line 475, in run_python_tests
    setup = await Get(
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 120, in __await__
    result = yield self
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 626, in native_engine_generator_send
    res = rule.send(arg) if err is None else rule.throw(throw or err)
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/backend/python/goals/pytest_runner.py", line 262, in setup_pytest_for_target
    ) = await MultiGet(
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 451, in MultiGet
    return await _MultiGet((__arg0, __arg1, __arg2, __arg3, __arg4, __arg5))
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 167, in __await__
    result = yield self.gets
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 626, in native_engine_generator_send
    res = rule.send(arg) if err is None else rule.throw(throw or err)
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/backend/python/goals/pytest_runner.py", line 262, in setup_pytest_for_target
    ) = await MultiGet(
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 451, in MultiGet
    return await _MultiGet((__arg0, __arg1, __arg2, __arg3, __arg4, __arg5))
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 167, in __await__
    result = yield self.gets
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 626, in native_engine_generator_send
    res = rule.send(arg) if err is None else rule.throw(throw or err)
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/backend/python/goals/pytest_runner.py", line 475, in run_python_tests
    setup = await Get(
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 120, in __await__
    result = yield self
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 626, in native_engine_generator_send
    res = rule.send(arg) if err is None else rule.throw(throw or err)
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/backend/python/goals/pytest_runner.py", line 262, in setup_pytest_for_target
    ) = await MultiGet(
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 451, in MultiGet
    return await _MultiGet((__arg0, __arg1, __arg2, __arg3, __arg4, __arg5))
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 167, in __await__
    result = yield self.gets
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 626, in native_engine_generator_send
    res = rule.send(arg) if err is None else rule.throw(throw or err)
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/backend/python/util_rules/pex.py", line 657, in create_pex
    result = await Get(BuildPexResult, PexRequest, request)
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 120, in __await__
    result = yield self
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 626, in native_engine_generator_send
    res = rule.send(arg) if err is None else rule.throw(throw or err)
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/backend/python/goals/pytest_runner.py", line 262, in setup_pytest_for_target
    ) = await MultiGet(
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 451, in MultiGet
    return await _MultiGet((__arg0, __arg1, __arg2, __arg3, __arg4, __arg5))
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 167, in __await__
    result = yield self.gets
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 626, in native_engine_generator_send
    res = rule.send(arg) if err is None else rule.throw(throw or err)
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/backend/python/util_rules/pex.py", line 657, in create_pex
    result = await Get(BuildPexResult, PexRequest, request)
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 120, in __await__
    result = yield self
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 626, in native_engine_generator_send
    res = rule.send(arg) if err is None else rule.throw(throw or err)
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/backend/python/util_rules/pex.py", line 657, in create_pex
    result = await Get(BuildPexResult, PexRequest, request)
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 120, in __await__
    result = yield self
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 626, in native_engine_generator_send
    res = rule.send(arg) if err is None else rule.throw(throw or err)
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/backend/python/util_rules/pex.py", line 657, in create_pex
    result = await Get(BuildPexResult, PexRequest, request)
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 120, in __await__
    result = yield self
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 626, in native_engine_generator_send
    res = rule.send(arg) if err is None else rule.throw(throw or err)
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/backend/python/util_rules/pex.py", line 657, in create_pex
    result = await Get(BuildPexResult, PexRequest, request)
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 120, in __await__
    result = yield self
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 626, in native_engine_generator_send
    res = rule.send(arg) if err is None else rule.throw(throw or err)
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/backend/python/util_rules/pex.py", line 657, in create_pex
    result = await Get(BuildPexResult, PexRequest, request)
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 120, in __await__
    result = yield self
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 626, in native_engine_generator_send
    res = rule.send(arg) if err is None else rule.throw(throw or err)
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/backend/python/util_rules/pex.py", line 531, in build_pex
    pex_python_setup = await _determine_pex_python_and_platforms(request)
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/backend/python/util_rules/pex.py", line 401, in _determine_pex_python_and_platforms
    python = await Get(
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 120, in __await__
    result = yield self
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 626, in native_engine_generator_send
    res = rule.send(arg) if err is None else rule.throw(throw or err)
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/backend/python/util_rules/pex.py", line 531, in build_pex
    pex_python_setup = await _determine_pex_python_and_platforms(request)
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/backend/python/util_rules/pex.py", line 401, in _determine_pex_python_and_platforms
    python = await Get(
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 120, in __await__
    result = yield self
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 626, in native_engine_generator_send
    res = rule.send(arg) if err is None else rule.throw(throw or err)
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/backend/python/util_rules/pex.py", line 531, in build_pex
    pex_python_setup = await _determine_pex_python_and_platforms(request)
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/backend/python/util_rules/pex.py", line 401, in _determine_pex_python_and_platforms
    python = await Get(
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 120, in __await__
    result = yield self
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 626, in native_engine_generator_send
    res = rule.send(arg) if err is None else rule.throw(throw or err)
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/backend/python/util_rules/pex.py", line 312, in find_interpreter
    result = await Get(
  File "/Users/me/Library/Caches/nce/a7a7e5c8d0659936773c836830336c0e3743f47588d74ecdce11e40b499a7003/bindings/venvs/2.16.0.dev5/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 120, in __await__
    result = yield self
pants.engine.process.ProcessExecutionFailure: Process 'Find interpreter for constraints: CPython>=3.8.*' failed with exit code 1.
stdout:

stderr:
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/runpy.py", line 187, in _run_module_as_main
    mod_name, mod_spec, code = _get_main_module_details(_Error)
  File "/usr/local/lib/python3.8/runpy.py", line 221, in _get_main_module_details
    return _get_module_details(main_name)
  File "/usr/local/lib/python3.8/runpy.py", line 155, in _get_module_details
    code = loader.get_code(mod_name)
  File "<frozen zipimport>", line 159, in get_code
  File "<frozen zipimport>", line 709, in _get_module_code
  File "<frozen zipimport>", line 536, in _get_data
FileNotFoundError: [Errno 2] No such file or directory: './pex'

```
</details>

<details>
  <summary>__run.sh</summary>

```
#!/bin/bash
# This command line should execute the same process as pants did internally.
export CPPFLAGS= LANG=C.UTF-8 LDFLAGS= PATH=$'/usr/local/bin:/usr/local/sbin:/usr/sbin:/usr/bin:/sbin:/bin' PEX_IGNORE_RCFILES=true PEX_PYTHON_PATH=$'/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin' PEX_ROOT=.cache/pex_root
cd /private/var/folders/8b/0w6lmjjx7fsdm_6hhcrjfdfw0000gq/T/pants-sandbox-2F5JHo
/usr/local/bin/python ./pex --tmpdir .tmp --pip-version 20.3.4-patched --python-path $'/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin' --interpreter-constraint $'CPython>=3.8.*' -- -c $'import hashlib, os, sys\n\npython = os.path.realpath(sys.executable)\nprint(python)\n\nhasher = hashlib.sha256()\nwith open(python, "rb") as fp:\n  for chunk in iter(lambda: fp.read(8192), b""):\n      hasher.update(chunk)\nprint(hasher.hexdigest())\n'
```
</details>

--- a/src/rust/engine/fs/store/src/lib.rs
+++ b/src/rust/engine/fs/store/src/lib.rs
@@ -39,11 +39,10 @@ pub use crate::snapshot_ops::{SnapshotOps, SubsetParams};
 
 use std::collections::{BTreeMap, BTreeSet, HashMap, HashSet};
 use std::fmt::{self, Debug, Display};
-use std::fs::hard_link;
 use std::fs::OpenOptions;
 use std::future::Future;
 use std::io::{self, Read, Seek, Write};
-use std::os::unix::fs::{symlink, OpenOptionsExt, PermissionsExt};
+use std::os::unix::fs::{OpenOptionsExt, PermissionsExt};
 use std::path::{Path, PathBuf};
 use std::sync::{Arc, Weak};
 use std::time::{Duration, Instant};
@@ -66,6 +65,11 @@ use protos::require_digest;
 use remexec::{ServerCapabilities, Tree};
 use serde_derive::Serialize;
 use sharded_lmdb::DEFAULT_LEASE_TIME;
+#[cfg(target_os = "macos")]
+use tokio::fs::copy;
+#[cfg(target_os = "linux")]
+use tokio::fs::hard_link;
+use tokio::fs::symlink;
 use tryfuture::try_future;
 use workunit_store::{in_workunit, Level, Metric};
 
@@ -73,13 +77,11 @@ const KILOBYTES: usize = 1024;
 const MEGABYTES: usize = 1024 * KILOBYTES;
 const GIGABYTES: usize = 1024 * MEGABYTES;
 
-/// How big a file must be to become an immutable input symlink.
+/// How big a file must be to become an immutable input hardlink.
 // NB: These numbers were chosen after micro-benchmarking the code on one machine at the time of
 // writing. They were chosen using a rough equation from the microbenchmarks that are optimized
 // for somewhere between 2 and 3 uses of the corresponding entry to "break even".
-// TODO: Temporarily disabled until #18162 and/or #18153 are resolved, since hardlinks do not play
-// well with Docker on macOS.
-const IMMUTABLE_FILE_SIZE_LIMIT: usize = usize::MAX - 1;
+const IMMUTABLE_FILE_SIZE_LIMIT: usize = 512 * KILOBYTES;
 
 mod local;
 #[cfg(test)]
@@ -1431,7 +1433,7 @@ impl Store {
     destination: PathBuf,
     target: String,
   ) -> Result<(), StoreError> {
-    symlink(target, destination)?;
+    symlink(target, destination).await?;
     Ok(())
   }
 
@@ -1440,7 +1442,16 @@ impl Store {
     destination: PathBuf,
     target: String,
   ) -> Result<(), StoreError> {
-    hard_link(target, destination)?;
+    // On macOS, copy uses a copy-on-write syscall (fclonefileat) which creates a disconnected
+    // clone. It is more defensive than a hardlink, but has the same requirement that the source
+    // and destination filesystem are the same.
+    //
+    // It also has the benefit of playing nicely with Docker for macOS file virtualization: see
+    // #18162.
+    #[cfg(target_os = "macos")]
+    copy(target, destination).await?;
+    #[cfg(target_os = "linux")]
+    hard_link(target, destination).await?;
     Ok(())
   }
 

--- a/src/rust/engine/fs/store/src/lib.rs
+++ b/src/rust/engine/fs/store/src/lib.rs
@@ -67,7 +67,7 @@ use serde_derive::Serialize;
 use sharded_lmdb::DEFAULT_LEASE_TIME;
 #[cfg(target_os = "macos")]
 use tokio::fs::copy;
-#[cfg(target_os = "linux")]
+#[cfg(not(target_os = "macos"))]
 use tokio::fs::hard_link;
 use tokio::fs::symlink;
 use tryfuture::try_future;
@@ -1450,7 +1450,7 @@ impl Store {
     // #18162.
     #[cfg(target_os = "macos")]
     copy(target, destination).await?;
-    #[cfg(target_os = "linux")]
+    #[cfg(not(target_os = "macos"))]
     hard_link(target, destination).await?;
     Ok(())
   }

